File: lib/llm/agent.rb
# frozen_string_literal: true

module LLM
  # Abstract agent protocol — the one and only execution contract in this layer.
  #
  # An "agent" is an autonomous harness: we hand it a prompt, a system-context
  # string, and a working directory, and it acts on its own — reads tasks.jsonl,
  # runs `bin/tasks`, edits files. Our code never parses its output for meaning;
  # it streams a transcript to the user and reloads the Store when the file
  # changes on disk. Every backend (Claude CLI, Hermes, later the Claude Agent
  # SDK / opencode / pi) implements this. There is deliberately no separate
  # "completion" protocol — we reach local models by putting a harness in front
  # of them, never by calling a bare model and coercing its text ourselves.
  #
  # Backends vary only along two axes this base hides:
  #   - transport: how we drive the harness. The subclasses here all spawn a CLI
  #     and reuse this machinery; an SDK-transport adapter would override start/
  #     pump/cancel while keeping the same surface.
  #   - model: which model the harness runs, passed through to #command.
  #
  # Two entry points share one #command:
  #   - async (TUI): #start, then multiplex #io with IO.select, drain via #pump,
  #     #cancel to stop. Output is captured into #output.
  #   - sync  (CLI): #run_sync spawns with inherited stdio so output streams
  #     straight to the terminal, and returns true on a clean exit.
  class Agent
    CANCEL_TERM_GRACE = 0.15
    CANCEL_KILL_GRACE = 0.15

    attr_reader :output, :io, :process_status

    # root:    working directory the harness runs in (where tasks.jsonl lives).
    # system:  fully-resolved system-context string (TASK_AGENT.md + file
    #          locations), or nil. Adapters inject it however their CLI allows.
    # command: binary name/path — lets config point at a non-default install.
    def initialize(root:, system: nil, command: nil, **_opts)
      @root = root
      @system = system.to_s.empty? ? nil : system
      @command = command.to_s.empty? ? self.class.default_command : command
      @output = +""
      @pid = nil
      @io = nil
      @process_status = nil
      @cancelled = false
    end

    # The binary this adapter spawns unless config overrides it.
    def self.default_command = raise NotImplementedError

    # Build the argv for a run. stream: true => emit a live transcript incl.
    # tool activity (TUI); false => final answer only (sync CLI). Subclasses
    # override; both entry points funnel through here so there is one source of
    # truth for how a backend is invoked.
    def command(prompt, model:, stream: true) = raise NotImplementedError

    # Is the backend usable right now? The base checks the binary is on PATH;
    # subclasses add reachability probes (e.g. Hermes pings its model endpoint).
    # Never raises — a dead backend returns false so the UI can flash, not crash.
    def available? = command_on_path?(@command)

    def running? = !@pid.nil?
    def cancelled? = @cancelled
    def exit_status = @process_status&.exitstatus
    def success? = !!@process_status&.success?

    def start(prompt, model:, stream: true)
      @output = +""
      @process_status = nil
      @cancelled = false
      r, w = IO.pipe
      # pgroup: true puts the child in its own process group so #cancel can TERM
      # the whole tree — these harnesses spawn tool subprocesses that would
      # otherwise be orphaned when we kill just the leader.
      @pid = Process.spawn(*command(prompt, model: model, stream: stream),
                           in: File::NULL, out: w, err: w, chdir: @root, pgroup: true)
      w.close
      @io = r
      self
    rescue StandardError
      w&.close unless w&.closed?
      r&.close unless r&.closed?
      @pid = nil
      @io = nil
      raise
    end

    # Drain available output. Returns :running or :done.
    def pump
      return :done unless @io
      # read_nonblock returns BINARY; force UTF-8 so downstream string ops
      # (wrap/truncate against UTF-8 literals) don't raise CompatibilityError.
      @output << @io.read_nonblock(65_536).force_encoding("UTF-8")
      :running
    rescue IO::WaitReadable
      :running
    rescue EOFError
      finish
    end

    def cancel
      @cancelled = true
      close_output
      return :done unless @pid

      # Negative pid = signal the whole process group started in #start. Some
      # harnesses (or their tool subprocesses) trap TERM, so cancellation must
      # not block the raw-terminal UI indefinitely in wait2. Give the group a
      # short graceful window, escalate to KILL, then detach as a last resort so
      # the child can still be reaped without holding the UI hostage.
      signal_group("TERM")
      return :done if reap_within(CANCEL_TERM_GRACE)

      signal_group("KILL")
      return :done if reap_within(CANCEL_KILL_GRACE)

      detach_unreaped_child
      :done
    end

    # Blocking run for the CLI: inherit stdio so the harness streams to the
    # terminal live, and return whether it exited cleanly. Defaults to the
    # non-streaming (final-answer) shape since there's no UI to animate.
    def run_sync(prompt, model:, stream: false)
      system(*command(prompt, model: model, stream: stream), chdir: @root)
    end

    private

    # PATH lookup without spawning a shell. An explicit path is checked directly.
    def command_on_path?(bin)
      return File.executable?(bin) if bin.include?("/")

      ENV.fetch("PATH", "").split(File::PATH_SEPARATOR).any? do |dir|
        File.executable?(File.join(dir, bin))
      end
    end

    def finish
      close_output
      if @pid
        _pid, @process_status = Process.wait2(@pid)
        @pid = nil
      end
      :done
    rescue Errno::ECHILD
      @pid = nil
      :done
    end

    def close_output
      @io&.close
      @io = nil
    end

    def signal_group(signal)
      Process.kill(signal, -@pid) if @pid
    rescue Errno::ESRCH
      nil
    end

    def reap_within(seconds)
      deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + seconds
      loop do
        waited_pid, status = Process.wait2(@pid, Process::WNOHANG)
        if waited_pid
          @process_status = status
          @pid = nil
          return true
        end

        remaining = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC)
        return false unless remaining.positive?

        IO.select(nil, nil, nil, [remaining, 0.01].min)
      end
    rescue Errno::ECHILD
      @pid = nil
      true
    end

    def detach_unreaped_child
      pid = @pid
      @pid = nil
      Process.detach(pid) if pid
    rescue Errno::ECHILD
      nil
    end
  end
end

File: go/internal/llm/agent.go
// Package llm is the agent adapter layer: one execution contract, a
// config-assembled registry, and the built-in harnesses. It is the Go
// counterpart of lib/llm/.
//
// An "agent" is an autonomous harness: we hand it a prompt, a system-context
// string, and a working directory, and it acts on its own — reads tasks.jsonl,
// runs `bin/tasks`, edits files. Our code never parses its output for meaning;
// it streams a transcript to the user and reloads the store when the file
// changes on disk. There is deliberately no separate "completion" protocol — we
// reach local models by putting a harness in front of them, never by calling a
// bare model and coercing its text ourselves.
//
// Backends vary along exactly two axes, and Agent hides both:
//
//   - transport: how the harness is driven. Every adapter here spawns a CLI and
//     reuses this machinery; an SDK-transport adapter would supply a different
//     Harness and keep the same Agent surface.
//   - model: which model the harness runs, passed through to Command.
//
// Two entry points share one Command, so there is one source of truth for how a
// backend is invoked:
//
//   - sync (CLI): RunSync spawns with inherited stdio so output streams straight
//     to the terminal, and reports whether it exited cleanly.
//   - async (TUI): Start, then read Output as it fills and wait on Done; Cancel
//     stops the whole process group.
package llm

import (
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"syscall"
	"time"
)

// Cancellation grace windows. Some harnesses (or the tool subprocesses they
// spawn) trap TERM, so cancellation must not block a raw-terminal UI
// indefinitely: the group gets a short graceful window, then KILL, then the
// child is detached rather than held onto.
const (
	CancelTermGrace = 150 * time.Millisecond
	CancelKillGrace = 150 * time.Millisecond
)

// Harness is the per-backend half of the contract, and the only thing an
// adapter implements. Everything else — spawning, availability, cancellation,
// output capture — is Agent's and is identical for every provider, which is what
// keeps provider-specific branching out of every call site.
type Harness interface {
	// DefaultBinary is the executable this adapter spawns unless config points
	// at another install.
	DefaultBinary() string
	// Argv builds the command line for one run. `system` is the fully resolved
	// system context, or "" when there is none; adapters inject it however their
	// CLI allows. `stream` true means "emit a live transcript including tool
	// activity" (the TUI); false means "final answer only" (the sync CLI).
	Argv(binary, system, prompt, model string, stream bool) []string
	// Reachable is any probe BEYOND "the binary is on PATH" — Hermes pings its
	// model endpoint. An adapter with nothing further to check returns true.
	// Never returns an error: a dead backend is unavailable, not a crash.
	Reachable() bool
}

// Options are what constructing an adapter needs. The model is not here: it
// rides along at call time, because one constructed agent can serve several
// models.
type Options struct {
	// Root is the working directory the harness runs in — where tasks.jsonl
	// lives, not where this binary lives.
	Root string
	// System is the resolved system-context string (TASK_AGENT.md + the file
	// locations + task-set memory), or "" for none.
	System string
	// Binary overrides the harness's default executable.
	Binary string
	// Path is the PATH the availability probe searches. It is passed in rather
	// than read from the process, because this layer must honour whatever
	// environment the adapter boundary resolved. Empty means "nothing is on
	// PATH", which is Ruby's ENV.fetch("PATH", "") exactly.
	Path string

	// OllamaURL and InferenceProvider are Hermes' settings; other adapters
	// ignore them. InferenceProvider is a pointer so that "set to empty" (drop
	// the --provider flag) is expressible and distinct from "unset" (use the
	// conventional local-Ollama provider name).
	OllamaURL         string
	InferenceProvider *string
}

// Agent is one constructed backend, ready to RunSync or Start.
type Agent struct {
	harness Harness
	root    string
	system  string
	binary  string
	path    string

	// mu guards everything below it. Output is read by a UI while the reader
	// goroutine appends to it, so the buffer is not merely convention-protected.
	mu        sync.Mutex
	output    []byte
	cmd       *exec.Cmd
	running   bool
	cancelled bool
	state     *os.ProcessState
	done      chan struct{}
}

// New constructs an agent over a harness. An empty Binary takes the harness's
// default, and an empty System means no system context at all — the same
// `system.to_s.empty? ? nil : system` normalisation Ruby applies, so a caller
// cannot accidentally inject a blank block.
func New(harness Harness, options Options) *Agent {
	binary := options.Binary
	if binary == "" {
		binary = harness.DefaultBinary()
	}
	closed := make(chan struct{})
	close(closed)
	return &Agent{
		harness: harness,
		root:    options.Root,
		system:  options.System,
		binary:  binary,
		path:    options.Path,
		done:    closed,
	}
}

// Binary is the executable this agent spawns.
func (a *Agent) Binary() string { return a.binary }

// System is the resolved system context, or "".
func (a *Agent) System() string { return a.system }

// Command is the argv for a run. Both entry points funnel through it.
func (a *Agent) Command(prompt, model string, stream bool) []string {
	return a.harness.Argv(a.binary, a.system, prompt, model, stream)
}

// Available answers whether the backend is usable right now: the binary is
// reachable, and the harness's own probe passes. It never returns an error — a
// dead backend is a false, so a UI can flash rather than crash.
func (a *Agent) Available() bool {
	return a.binaryOnPath() && a.harness.Reachable()
}

// binaryOnPath is Agent#command_on_path?: an explicit path is checked directly,
// and a bare name is searched along PATH without spawning a shell.
func (a *Agent) binaryOnPath() bool {
	if strings.Contains(a.binary, "/") {
		return executable(a.binary)
	}
	for _, dir := range strings.Split(a.path, string(os.PathListSeparator)) {
		if executable(filepath.Join(dir, a.binary)) {
			return true
		}
	}
	return false
}

// executable is File.executable?: a regular file (or a symlink to one) carrying
// an execute bit. A directory named like the binary is not executable, which is
// the one case a naive mode check gets wrong.
func executable(path string) bool {
	info, err := os.Stat(path)
	if err != nil || info.IsDir() {
		return false
	}
	return info.Mode().Perm()&0o111 != 0
}

// RunSync is the blocking run for the CLI: inherit stdio so the harness streams
// to the terminal live, and report whether it exited cleanly. It defaults to the
// non-streaming (final-answer) shape at the call site, since there is no UI to
// animate.
//
// The boolean is Kernel#system's: true on a clean exit, false on any nonzero
// status AND on a spawn that never happened. Both mean "the agent did not
// complete", which is the only distinction the caller acts on.
func (a *Agent) RunSync(prompt, model string, stream bool) bool {
	argv := a.Command(prompt, model, stream)
	if len(argv) == 0 {
		return false
	}
	command := exec.Command(argv[0], argv[1:]...)
	command.Dir = a.root
	command.Env = childEnviron()
	command.Stdin = os.Stdin
	command.Stdout = os.Stdout
	command.Stderr = os.Stderr
	return command.Run() == nil
}

// childEnviron is the environment a harness inherits: this process's, unchanged.
//
// It is set EXPLICITLY rather than left nil, and that is the whole point. With a
// nil Env and a non-empty Dir, os/exec appends `PWD=<Dir>` — a variable Ruby's
// `system(*argv, chdir:)` never sets. A differential run caught it: the same
// recorder script printed `/var/folders/…` under Go and `/private/var/folders/…`
// under Ruby, because sh's `pwd` prefers an exported PWD over the physical path.
// The directory was the same one either way, but the harness's environment was
// not, and a harness is an external contract we do not get to quietly amend.
func childEnviron() []string { return os.Environ() }

// Start spawns the harness asynchronously, capturing its combined output. The
// child is put in its own process group so Cancel can signal the whole tree:
// these harnesses spawn tool subprocesses that would otherwise be orphaned when
// only the leader is killed.
//
// This is the async surface Ruby's Agent#start/#pump/#cancel provides, in the
// shape Go gives it: a goroutine drains the pipe into Output and closes Done,
// rather than a caller multiplexing with IO.select. The observable contract —
// output captured, exit status recorded, cancellation distinct from a nonzero
// exit, a fresh run resetting the metadata — is the same.
func (a *Agent) Start(prompt, model string, stream bool) error {
	argv := a.Command(prompt, model, stream)
	if len(argv) == 0 {
		return errors.New("llm: adapter produced an empty command")
	}

	reader, writer, err := os.Pipe()
	if err != nil {
		return err
	}
	command := exec.Command(argv[0], argv[1:]...)
	command.Dir = a.root
	command.Env = childEnviron()
	command.Stdin = nil // File::NULL: a headless harness must never block on input.
	command.Stdout = writer
	command.Stderr = writer
	command.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}

	a.mu.Lock()
	defer a.mu.Unlock()
	if err := command.Start(); err != nil {
		reader.Close()
		writer.Close()
		return err
	}
	// Our copy of the write end must go once the child holds its own, or the
	// reader never sees EOF and Done never closes.
	writer.Close()

	a.output = nil
	a.state = nil
	a.cancelled = false
	a.running = true
	a.cmd = command
	done := make(chan struct{})
	a.done = done

	go a.drain(command, reader, done)
	return nil
}

// drain copies the child's output until EOF, reaps it, and closes done.
func (a *Agent) drain(command *exec.Cmd, reader *os.File, done chan struct{}) {
	buffer := make([]byte, 65536)
	for {
		n, err := reader.Read(buffer)
		if n > 0 {
			a.mu.Lock()
			a.output = append(a.output, buffer[:n]...)
			a.mu.Unlock()
		}
		if err != nil {
			break
		}
	}
	reader.Close()
	waitErr := command.Wait()

	a.mu.Lock()
	if command.ProcessState != nil {
		a.state = command.ProcessState
	}
	a.running = false
	a.cmd = nil
	a.mu.Unlock()
	_ = waitErr
	close(done)
}

// Done closes when the run has exited and its output has been drained. A fresh
// Agent's channel is already closed, so a caller that selects on it before any
// run does not block forever.
func (a *Agent) Done() <-chan struct{} {
	a.mu.Lock()
	defer a.mu.Unlock()
	return a.done
}

// Output is the transcript captured so far. It is a snapshot: safe to call while
// the run is still producing.
func (a *Agent) Output() string {
	a.mu.Lock()
	defer a.mu.Unlock()
	return string(a.output)
}

func (a *Agent) Running() bool {
	a.mu.Lock()
	defer a.mu.Unlock()
	return a.running
}

// Cancelled reports whether this run was stopped by Cancel. It is deliberately
// distinct from Success: a cancelled run and a run that failed on its own both
// end nonzero, and only the flag tells a UI which sentence to show.
func (a *Agent) Cancelled() bool {
	a.mu.Lock()
	defer a.mu.Unlock()
	return a.cancelled
}

// ExitStatus is the child's exit code, or -1 when it has not exited or died on a
// signal.
func (a *Agent) ExitStatus() int {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.state == nil {
		return -1
	}
	return a.state.ExitCode()
}

func (a *Agent) Success() bool {
	a.mu.Lock()
	defer a.mu.Unlock()
	return a.state != nil && a.state.Success()
}

// Signaled reports whether the child died on a signal, and which one. This is
// how a caller distinguishes "the harness refused to stop and we killed it"
// from "the harness exited".
func (a *Agent) Signaled() (bool, syscall.Signal) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.state == nil {
		return false, 0
	}
	status, ok := a.state.Sys().(syscall.WaitStatus)
	if !ok || !status.Signaled() {
		return false, 0
	}
	return true, status.Signal()
}

// Cancel stops a running agent. The signal goes to the whole process group
// started in Start — negative pid — so a harness's tool subprocesses go with it.
//
// TERM first, then KILL after a short window, and if even that is not reaped in
// time the child is left to the drain goroutine rather than blocking the caller.
// A UI must never be held hostage by a harness that traps signals.
func (a *Agent) Cancel() {
	a.mu.Lock()
	a.cancelled = true
	command, done := a.cmd, a.done
	a.mu.Unlock()

	if command == nil || command.Process == nil {
		return
	}
	signalGroup(command.Process.Pid, syscall.SIGTERM)
	if waitFor(done, CancelTermGrace) {
		return
	}
	signalGroup(command.Process.Pid, syscall.SIGKILL)
	waitFor(done, CancelKillGrace)
}

func signalGroup(pid int, signal syscall.Signal) {
	// A negative pid addresses the group. ESRCH means it is already gone, which
	// is the outcome we wanted.
	_ = syscall.Kill(-pid, signal)
}

func waitFor(done <-chan struct{}, grace time.Duration) bool {
	timer := time.NewTimer(grace)
	defer timer.Stop()
	select {
	case <-done:
		return true
	case <-timer.C:
		return false
	}
}
